### PR TITLE
Implement common definition file for multi api classes

### LINKED ISSUE #987: [Bug] - Mirroring External Types with multiple files - generates Redundant types in the dart code
### Describe the bug

Mirroring External Types with multiple files - generates Redundant types in the dart code

**my_types.rs**
```rs
#[frb(mirror(Configuration))]
#[derive(Default, Clone)]
pub struct _Configuration {
    pub id: u32,
    pub name: String
}
```

**api1.rs**
```rs
pub fn update_configuration(config: Configuration) -> anyhow::Result<()> {
   
}
```

**api2.rs**
```rs
pub fn perform_some_operaton(config: Configuration) -> anyhow::Result<()> {
   
}
```

```rs
// build.rs contains all rust input and output files
cargo build
```

### Codegen logs with `RUST_LOG=debug` environment variable

```shell
Finished dev [unoptimized + debuginfo] target(s) in 10.40s
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Generated binding code

Api1.dart

```dart
abstract Api1Class {
 . . .
}
 class Configuration {     <----- this class is generated in both Api1 and Api2
   . . .
 } 

class Api1ClassImpl implements Api1Class {
 
   . . . 
  Future<void> updateConfiguration (Configuration config) {

  }
  
}
```

**Api2.dart**
```dart
abstract Api2Class {
 . . .
}
 class Configuration {     <----- this class is generated in both Api1 and Api2
   . . .
 } 

class Api2ClassImpl implements Api2Class {
 
   . . . 
  Future<void> performSomeOtherAction(Configuration config) {
    . . .
  }
   
  
}
```

### OS

MacOS

### Version of `flutter_rust_bridge_codegen`

1.61.*

### Flutter info

_No response_

### Version of `clang++`

_No response_

### Version of `ffigen`

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by codercengiz**
## Changes

Fixes #987 

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (via `just precommit`).
- [ ] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [ ] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Good job!

Firstly, may I know whether this is related to https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145?

**Comment by codercengiz** [comment]
> Good job!
> 
> Firstly, may I know whether this is related to #1145?

Yes, it is related to common structs in case multi-blocks (multi-api)

**Comment by fzyzcjy** [comment]
I will check this in more detail later since quite busy now. Anyway just want to show you #1145 to avoid this PR and #1145 are in conflict :)

**Comment by codercengiz** [comment]
Just more info:

Creating a separate dart definition file for each external library might be a better approach, but it requires making a lot of changes to identify which structs are external and even to determine which struct belongs to which external library if multiple external libraries are used. That's why I chose this method. By doing so, I collected all the used structs in a single definition file regardless of whether they are external or not.

**Comment by fzyzcjy** [comment]
(Quite busy recently, I do hope I can have time today reviewing this!)

**Comment by fzyzcjy** [comment]
Start reviewing now.

Firstly, cc @dbsxdbsx who implemented the initial multi-file feature - do you have an interest to take a look / review this approach :)

File: justfile
Comment (fzyzcjy) [inline justfile:125]: nit: So, shall we use `pure_dart_multi` as a test bed for this PR?

File: frb_codegen/src/lib.rs
Comment (fzyzcjy) [inline frb_codegen/src/lib.rs:82]: nit: shall this be a `pub` fn (i.e. exposed as public API)?

**Review by fzyzcjy — COMMENTED** [review]
(partially reviewed; will continue later)

**Review by fzyzcjy — COMMENTED** [review]
(partially reviewed; will continue later)

**Comment by fzyzcjy** [comment]
I realized https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145 says #987 will be closed by that PR, while this PR also wants to close the issue. Therefore, as I mentioned before, I wonder whether the two PRs will be in conflict? They both seem to solve the conflicts-when-multi-file problem. If so, feel free to cooperate!

**Comment by dbsxdbsx** [comment]
> Start reviewing now.
> 
> Firstly, cc @dbsxdbsx who implemented the initial multi-file feature - do you have an interest to take a look / review this approach :)

Frankly, no time at present (and for the coming 3 days). I am ok to merge this pr even if it has conflict with my pr.

At present, in my pr https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145, it is not time yet to deal with the shared stuff with `dart-decl-output`.  But it is for sure this flag would be taken into account soon.

**Comment by fzyzcjy** [comment]
@dbsxdbsx Take your time and looking forward to your PR! So if I understand correctly, I should continue reviewing this PR (and possibly merge it if everything is done).

**Comment by dbsxdbsx** [comment]
> @dbsxdbsx Take your time and looking forward to your PR! So if I understand correctly, I should continue reviewing this PR (and possibly merge it if everything is done).

Yes, since I can't give any constructive idea on this flag now. So, even if the design in this pr is conflict with mine, I would try to solve it then.

**Comment by codercengiz** [comment]
@dbsxdbsx  I absolutely agree. We need a permanent and elegant solution for multi-API. That's why I'm eagerly waiting for #1145 PR to be finished. Although my solution in this PR is a quick and simple one, it may not be able to address future requests. But as I mentioned earlier, taking a fresh approach to the multi-API structure is a long-term strategy, and I'm not very familiar with this project, so I suggested a simple solution.

If @dbsxdbsx  is also okay with it, we can merge this PR, but we still need the new mechanism in #1145 anyway.

File: frb_codegen/src/lib.rs
Comment (codercengiz) [reply frb_codegen/src/lib.rs:82]: It should be accessible from `main.rs`, so I set it as public. 

The related function should find common structs by looking at all `IrFile` at once, not each one individually. That's why I thought it should be called in `main.rs`.

File: justfile
Comment (codercengiz) [reply justfile:125]: "We can use the `pure_dart_multi` example. However, since `dart-decl-output` is not set in this example, we won't be able to see the changes in this PR and won't be able to catch any future errors. Maybe we should add a fourth example called `pure_dart_multi_with_decl_output`. 

However, I still prefer having more examples on the Flutter side. So we can add the fourth example on the Flutter side, like `with_flutter_multi`."

**Comment by fzyzcjy** [comment]
Take your time! I am also quite busy recently, and since this PR is a bit nontrivial I have to find a slot of time so need to wait for a bit ;)

File: frb_codegen/src/lib.rs
Comment (fzyzcjy) [reply frb_codegen/src/lib.rs:82]: Ok I see. Hopefully a later refactor can unify many things :)

File: justfile
Comment (fzyzcjy) [reply justfile:125]: As for more testing examples, possibly related to: https://github.com/fzyzcjy/flutter_rust_bridge/issues/1210

File: frb_codegen/src/entrypoint/dart.rs
Comment (fzyzcjy) [inline frb_codegen/src/entrypoint/dart.rs:156]: nit: Are we guaranteed that `dart_decl_output_path` exists and only exists in configs[0]? If so we can add an sanity check assertion

File: frb_example/with_flutter/rust/src/api2.rs
Comment (fzyzcjy) [inline frb_example/with_flutter/rust/src/api2.rs:1]: nit: would be great to add comments explaining we are *deliberately* repeating ourselves here. Otherwise it looks very like we are accidentically copy-pasting code

**Review by fzyzcjy — COMMENTED** [review]
Finished a brief review. Too tired today so not sure I am missing something or not. But anyway good job and only nits!

File: frb_codegen/src/entrypoint/dart.rs
Comment (codercengiz) [reply frb_codegen/src/entrypoint/dart.rs:157]: We guaranteed that dart_decl_output_path exists, but I added the sanity check assertion anyway.

**Comment by fzyzcjy** [comment]
(I am not sure whether the PR is ready for review again - ping me when needed!)

**Comment by codercengiz** [comment]
> (I am not sure whether the PR is ready for review again - ping me when needed!)
@fzyzcjy 
Yes it is ready to review again

**Comment by fzyzcjy** [comment]
Great! I will check as soon as possible, since still quite busy

**Comment by dbsxdbsx** [comment]
@fzyzcjy ,fot github Action `CI / Flutter integration test on MacOS`, I found it always failed (not only in my [pr](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145)) after merging https://github.com/fzyzcjy/flutter_rust_bridge/pull/1213(updating flutter to 3.10).  Meanwhile, the issue log is not clear(especially for  those who don't have MacOs)
Does it become a  persistent issue?

**Comment by fzyzcjy** [comment]
@dbsxdbsx Just ignore it now, see https://github.com/fzyzcjy/flutter_rust_bridge/issues/1225 for details

File: frb_codegen/src/entrypoint/dart.rs
Comment (fzyzcjy) [inline frb_codegen/src/entrypoint/dart.rs:1]: (unrelated to this file) will we introduce breaking changes to APIs?

**Review by fzyzcjy — APPROVED** [review]
Good job!

**Comment by fzyzcjy** [comment]
To be honest, I may not think it is the best place to put those common structs, but changing it may require more PRs. Given @dbsxdbsx's opinions, who implements the original multi-file feature, I will firstly merge this feature, and later wait for @dbsxdbsx's further PRs :)

**Comment by svenstaro** [comment]
Is there any chance this could be merged? The current limitations are blocking our use case.

File: frb_codegen/src/entrypoint/dart.rs
Comment (codercengiz) [reply frb_codegen/src/entrypoint/dart.rs:1]: There is no breaking change

**Comment by fzyzcjy** [comment]
@svenstaro I am willing to merge after CI goes green!

**Comment by codercengiz** [comment]
@fzyzcjy 
After the changes in PR #1285, we can no longer use the same external struct in multiple APIs. The `into_into_dart` trait, as seen below, causes a conflict error. This error is also present in the master branch.
image

Two different trait functions are being generated for the same struct, causing the Rust compiler to be unable to decide which one to use. If I manually delete the trait functions generated from either `bridge_generated_api.rs` or `bridge_generated_api2.rs`, the error is resolved.

A change needs to be made in the Rust code generator part. However, this PR only contains changes in the Dart code generator. Therefore, I'm not sure, but I think this should be addressed in a separate issue.

**Comment by fzyzcjy** [comment]
Hmm that looks like a problem... That's why it is often a good idea to merge PRs as soon as possible (but need to be after CI is green etc)...

So, will @dbsxdbsx's PR be helpful for this scenario?

**Comment by dbsxdbsx** [comment]
> Hmm that looks like a problem... That's why it is often a good idea to merge PRs as soon as possible (but need to be after CI is green etc)...
> 
> So, will @dbsxdbsx's PR be helpful for this scenario?

I didn't look into this pr, but I guess this pr would be deprecated since mine changed lots of things,

**Comment by fzyzcjy** [comment]
> I didn't look into this pr, but for the issue into_into_dart trait, I did handle it before in my pr-- yes, it is TOUGH to make it consistent with the latest frb. And I mean no offense, but I guess this pr would be deprecated since mine changed many things and would be more consistent with the latest frb.

Ah that does happen sometimes... concurrent development makes some PR deprecated. That's life :/

**Comment by svenstaro** [comment]
@fzyzcjy can you update us on where we are with this in the grand scheme of things? I see that we PR won't be merged but what's the current idea for flutter-rust-bridge to deal with the problem outlined in this PR?

**Comment by fzyzcjy** [comment]
@svenstaro Hi, @dbsxdbsx has been doing a PR https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145 that enhances the multi-file scenario, and I am eagerly looking forward to merging that PR!

**Comment by dbsxdbsx** [comment]
> @svenstaro Hi, @dbsxdbsx has been doing a PR #1145 that enhances the multi-file scenario, and I am eagerly looking forward to merging that PR!

I just happened to give a temp [report](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145#issuecomment-1762637641), hope my pr could be solved ASAP.

**Comment by svenstaro** [comment]
Thanks guys, really looking forward to that. Can you close this PR then?

**Comment by fzyzcjy** [comment]
This is implemented in flutter_rust_bridge v2. Thank you for your previous attempts and efforts!

**Comment by dbsxdbsx** [comment]
> This is implemented in flutter_rust_bridge v2. Thank you for your previous attempts and efforts!

Nice to hear. I guess 1.82.6 is just a temporary version? Then when will the v2.0 stable version come?

**Comment by fzyzcjy** [comment]
Sure! 1.82.6 is only there to ensure that, when 2.0.0-dev.12345 etc publish, people on pub.dev will see 1.x readme by default, and I want them to know 2.x is there :)

**Pull request closed without merging**
